src/main/java/runhistoryplus/patches/PotionUseAndDiscardRunHistoryPatch.java

```diff
@@ -179,4 +251,98 @@ public int[] Locate(CtBehavior ctMethodToPatch) throws CannotCompileException, P
+                    PotionsObtainedEntropicBrewLog.potions_obtained_entropic_brew.get(PotionsObtainedEntropicBrewLog.potions_obtained_entropic_brew.size() - 1).add(potion.ID);
+                }
+                break;
+        }
+    }
+
+    @SpirePatch(clz = ObtainPotionAction.class, method = "update")
+    public static class ObtainPotionActionAddLogging {
+        public static class ObtainPotionActionAddLoggingExprEditor extends ExprEditor {
+            @Override
+            public void edit(MethodCall methodCall) throws CannotCompileException {
+                if (methodCall.getMethodName().equals("obtainPotion")) {
+                    methodCall.replace(String.format("{ $_ = $proceed($$); if ($_) %1$s.obtainPotionAlchemizeOrEntropicBrew(this.potion, (String)%2$s.sourceField.get(this)); }",
+                            PotionUseAndDiscardRunHistoryPatch.class.getName(), ActionSourceField.class.getName()));
+                }
+            }
+        }
+
+        @SpireInstrumentPatch
+        public static ExprEditor obtainPotionActionAddLoggingPatch() {
+            return new ObtainPotionActionAddLoggingExprEditor();
+        }
+    }
+
+    @SpirePatch(clz = ObtainPotionEffect.class, method = "update")
+    public static class ObtainPotionEffectAddLogging {
+        public static class ObtainPotionEffectAddLoggingExprEditor extends ExprEditor {
+            @Override
+            public void edit(MethodCall methodCall) throws CannotCompileException {
+                if (methodCall.getMethodName().equals("obtainPotion")) {
+                    methodCall.replace(String.format("{ $_ = $proceed($$); if ($_) %1$s.obtainPotionAlchemizeOrEntropicBrew(this.potion, (String)%2$s.sourceField.get(this)); }",
+                            PotionUseAndDiscardRunHistoryPatch.class.getName(), EffectSourceField.class.getName()));
+                }
+            }
+        }
+
+        @SpireInstrumentPatch
+        public static ExprEditor obtainPotionEffectAddLoggingPatch() {
+            return new ObtainPotionEffectAddLoggingExprEditor();
+        }
+    }
+
+    @SpirePatch(clz = ObtainPotionAction.class, method = SpirePatch.CLASS)
+    public static class ActionSourceField {
+        public static final SpireField<String> sourceField = new SpireField<>(() -> "");
+    }
+
+    @SpirePatch(clz = ObtainPotionEffect.class, method = SpirePatch.CLASS)
+    public static class EffectSourceField {
+        public static final SpireField<String> sourceField = new SpireField<>(() -> "");
+    }
+
+    @SpirePatch(clz = Alchemize.class, method = "use")
+    public static class AlchemizeExprEditor extends ExprEditor {
+        @Override
+        public void edit(NewExpr newExpr) throws CannotCompileException {
+            if (newExpr.getClassName().equals(ObtainPotionAction.class.getName())) {
+                newExpr.replace(String.format("{ $_ = $proceed($$); %1$s.sourceField.set($_, \"Alchemize\"); }", ActionSourceField.class.getName()));
```

Comment:
Let's make these two values (Alchemize here and Entropic Brew below) into constants or an enum that we use here and in `obtainPotionAlchemizeOrEntropicBrew`. Doing this makes sure that the connection between the different places that set or check this is enforced by the code.

I think the enum version works best (and I'll put that out below), but let me start by writing out the version using constants, since it's like an intermediate step between what the code looks like now and the enum version. So here's a rough sketch of what it would look like using constants:
```
//at the top of this file, right after the static logger
public static final String ALCHEMIZE = "Alchemize";
...
// in obtainPotionAlchemizeOrEntropicBrew
case ALCHEMIZE:
...
//in the patch here
newExpr.replace(String.format("{ $_ = $proceed($$); %1$s.sourceField.set($_, \"%2$s\"); }", ActionSourceField.class.getName(), ALCHEMIZE));
```

Now, with an enum, this would look like:
```
//this definition would go in PotionUseAndDiscardRunHistoryPatch
public enum PotionSource {
	Alchemize,
	EntropicBrew
}
...
//then we'd change the field to use this enum instead, and to default to null
@SpirePatch(clz = ObtainPotionAction.class, method = SpirePatch.CLASS)
public static class ActionSourceField {
    public static final SpireField<PotionSource> sourceField = new SpireField<>(() -> null);
}
...
// in obtainPotionAlchemizeOrEntropicBrew
case Alchemize:
...
//in the patch here (didn't actually test this one but it should be close to correct)
newExpr.replace(String.format("{ $_ = $proceed($$); %1$s.sourceField.set($_, %2$s.PotionSource.Alchemize); }", ActionSourceField.class.getName(), PotionUseAndDiscardRunHistoryPatch.class.getName()));
...
//etc. We would need to make minor changes to the other patches and functions that support logging this data but nothing too major (this is a rough sketch and not complete)
```

I see the enum version is most maintainable in the long run. The differences compared to the string constant version aren't that big... but enums are practically built for this use case, where you have a limited number of possibilities that you want to _enumerate_ (and you don't actually want any random string to be valid). It also makes it very easy to understand all the places the enum is used (and lets you do some other nice stuff that doesn't really matter for this work, but that adds to the reasons I feel enums are the generally preferred approach).

Hopefully that makes sense -- it's not a huge thing, but I prefer to follow best practices like this when it's reasonable to do so.

Reply:
You are of course absolutely correct and I'm a little bit embarrassed, that I didn't think of using an enum in the first place :sweat_smile: